src/ray/raylet/node_manager.cc

```diff
@@ -2897,25 +2897,52 @@ MemoryUsageRefreshCallback NodeManager::CreateMemoryUsageRefreshCallback() {
             << "worker pid: " << high_memory_eviction_target_->GetProcess().GetId()
             << "task: " << high_memory_eviction_target_->GetAssignedTaskId();
       } else {
-        auto newest_task_start_time = std::chrono::high_resolution_clock::time_point();
-        std::shared_ptr<WorkerInterface> latest_worker = nullptr;
-        for (const auto &worker : worker_pool_.GetAllRegisteredWorkers()) {
-          if (worker->GetAssignedTaskTime() > newest_task_start_time) {
-            newest_task_start_time = worker->GetAssignedTaskTime();
-            latest_worker = worker;
+        auto workers = worker_pool_.GetAllRegisteredWorkers();
+        if (!workers.empty()) {
+          std::sort(workers.begin(),
+                    workers.end(),
+                    [](std::shared_ptr<WorkerInterface> const &left,
+                       std::shared_ptr<WorkerInterface> const &right) -> bool {
+                      return left->GetAssignedTaskTime() > right->GetAssignedTaskTime();
+                    });
+          std::shared_ptr<WorkerInterface> latest_worker = workers.front();
+          std::ostringstream top_n_latest_workers_info;
+          int64_t top_n = 0;
+
+          const static int64_t max_to_print = 10;
+          for (auto &worker : workers) {
+            auto pid = worker->GetProcess().GetId();
+            auto used_memory = this->memory_monitor_->GetProcessMemoryBytes(pid);
+            auto task_assign_time =
+                std::chrono::system_clock::to_time_t(worker->GetAssignedTaskTime());
+            top_n_latest_workers_info
+                << "Worker " << top_n << " task assigned time "
+                << std::put_time(std::gmtime(&task_assign_time), "%c %Z")
+                << " memory used " << used_memory << " task spec "
+                << worker->GetAssignedTask().GetTaskSpecification().DebugString() << "\n";
+            top_n += 1;
+            if (top_n >= max_to_print) {
+              break;
+            }
           }
-        }
-        if (latest_worker != nullptr) {
-          RAY_LOG(INFO) << "Killing worker with the newest started task "
-                        << "to free up memory. "
-                        << "worker pid: " << latest_worker->GetProcess().GetId()
-                        << "task: " << latest_worker->GetAssignedTaskId();
+          RAY_LOG_EVERY_N_OR(ERROR, 10, INFO)
+              << "Killing worker with the latest task assigned time "
+              << "to free up memory. Task to be killed:"
```

Comment:
Can you mention to use `ray list workers --detail` to see workers that are killed by OOM killer?